treematcher.py

```diff
@@ -1,286 +1,178 @@
-from itertools import permutations
-from ete3 import NCBITaxa
-import six
+#system modules
 import re
-import ast
+from itertools import permutations
 
+# third party modules
+import ast
+import six
 from ete3 import PhyloTree, Tree, NCBITaxa
 
+# internal modules
+# ...
 
-class TreePattern(Tree):
-    """
-        Class for presenting a tree of constrained patterns. A way to represent patterns to search for within trees.
-        :param args: Pattern to be searched.
-        :param kargs: Custom functions and arguments passed to Tree Class.
-    """
-    _syntax_tuples = [
-        ("@", "__target"),
-        ("__target.size", "len(__target.get_leaves())"),
-        ("__target.leaves", "[node.name for node in __target.get_leaves()]"),
-        ("__target.contains_species", "[n.species for n in __target.traverse()]"),
-    ]
-
-    _syntax_tuples_with_cache = [
-        ("@", "__target"),
-        ("__target.size", "len(temp_leaf_cache[__target])"),
-        ("__target.contains_species", "[n.species for n in temp_leaf_cache[__target]]"),
-        ("__target.leaves", "[n.name for n in temp_leaf_cache[__target]]"),
-
-    ]
+class TreePatternCache(object):
+    def __init__(self, tree):
+        """ Creates a cache for attributes that require multiple tree
+        traversal when using complex TreePattern queries.
 
-    def __init__(self, *args, **kargs):
-        kargs["format"] = 1
-
-        is_exact = False
-
-        self.all_node_cache = None
-        self.shortcut_functions = []
-
-
-
-        try:
-            self._extra_functions = kargs.pop('functions')
+        :param tree: a regular ETE tree instance
+         """
+        # Initialize cache (add more stuff as needed)
+        self.leaves_cache = tree.get_cached_content()
+        self.all_node_cache = tree.get_cached_content(leaves_only=False)
```

Comment:
only this two caches are currently necessary. But we should include here anything else needed, for instance node2node distance matrix, the ncbiTaxa instance, etc.. 
Perhaps adding some arguments to __init__ where users can choose what to cache.